sdk/src/azure/core/az_http_request.c

```diff
@@ -107,11 +173,54 @@ az_http_request_set_query_parameter(_az_http_request* ref_request, az_span name,
   // name or value can't be empty
   _az_PRECONDITION(az_span_size(name) > 0 && az_span_size(value) > 0);
 
-  int32_t required_length = az_span_size(name) + az_span_size(value) + 2;
-
+  // check if query parameter is already in url
+  int32_t pre_existing_query_parameter_start_index = 0;
+  az_span pre_existing_query_parameter_value = _az_http_request_find_query_parameter(
+      ref_request, name, &pre_existing_query_parameter_start_index);
   az_span url_remainder
       = az_span_slice_to_end(ref_request->_internal.url, ref_request->_internal.url_length);
 
+  if (pre_existing_query_parameter_start_index > 0)
+  { // a negative difference means shifting left and no required length.
+    // a positive difference means new value will require shifting right.
+    int32_t pre_existing_query_parameter_value_size
+        = az_span_size(pre_existing_query_parameter_value);
+    int32_t difference = az_span_size(value) - pre_existing_query_parameter_value_size;
+    int32_t required_length = difference > 0 ? difference : 0;
+    AZ_RETURN_IF_NOT_ENOUGH_SIZE(url_remainder, required_length);
+
+    // There is already a query parameter name. update the value if different
+    if (az_span_is_content_equal(value, pre_existing_query_parameter_value))
+    {
+      // No need to do anything, existing value is equal to the new one
+      return AZ_OK;
+    }
+
+    // Replace the value content. This might shift right or left the url contents.
+    // No need to check
+    AZ_RETURN_IF_FAILED(_az_span_replace(
+        ref_request->_internal.url,
+        ref_request->_internal.url_length,
+        pre_existing_query_parameter_start_index,
+        pre_existing_query_parameter_start_index + pre_existing_query_parameter_value_size,
+        value));
+    ref_request->_internal.url_length += difference;
+    if (difference < 0)
+    {
+      // clear anything after a left shift
+      difference *= -1;
+      az_span_fill(
+          az_span_slice(
+              az_span_slice_to_end(ref_request->_internal.url, ref_request->_internal.url_length),
+              0,
+              difference),
```

Comment:
Can we clear everything beyond the end of `url_length`? Why do we slice from `0 to difference`?

Reply:
performance. 
If we have a very huge url buffer, and shift left just one position multiple times, we would be filling and filling all the remaining bytes over and over.
That's why we just care about the content that was left shifted

Reply:
I see. In that case, is it possible for us to be in a state where query parameters/data from older requests (when using a shared array) persists? Or, since every set and shift left will clear the "remaining", will we never have old data in the buffer, beyond the url length?

Imagine you have this in your buffer, as the first request:
`?q1=123&q2=456`

Then you create a second request:
`?q1=123` (&q2=456 is still there in the buffer)
You set the `q1` to a smaller length value, causing a left shift.
`?q1=1\0\0` (&q2=456 is still there in the buffer)
We cleared the bytes `23`, but what about the data from the first request? That is still visible. I can see the value of `q2`.

Reply:
yeah, I considered that one. But I am considering that all this method should care is about the leaking it might cause only.
The case you shared is not really this method fault right?
Don't we have the same case by re-using the shared array without even calling this method?  If an initial URL has queryparameters, and we never call set_query_(...), then we use the same buffer for a new request, we would see all query data in the url remainder span... And that's not `az_http_request_set_query_parameter` fault.

So, if we want to handle a case like that, we should not use `az_http_request_set_query_parameter` to clean, since it might be never called.
We would need to come up with another strategy, like using retry policy to clean query parameters upon completing a request, or something like that.

Also, consider that Requests are not exposed to public.  Request is internal and only used by SDK clients.  Each SDK client uses stack for the Request (url and headers). So, probably we should care more about the Response.